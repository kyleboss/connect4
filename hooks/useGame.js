import { useContext } from "react";

import GameContext from "@/state/GameContext";

const useGame = () => useContext(GameContext);

export default useGame;
