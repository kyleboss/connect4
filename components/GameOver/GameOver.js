import { useMemo } from "react";

import useGame from "@/hooks/useGame";
import { USER_ROLE_GUEST, USER_ROLES } from "@/utils/constants";
import Button from "@/components/Button";
import Header from "@/components/Header";
import Subheader from "@/components/Subheader";

const GameOver = () => {
  const { userRole, winner, resetGame } = useGame();

  const header = useMemo(
    () => (userRole === winner ? "Congratulations! ðŸŽ‰" : "Game over."),
    [userRole, winner]
  );

  const subheader = useMemo(
    () => (userRole === winner ? "You won." : `${USER_ROLES[winner]} won.`),
    [userRole, winner]
  );

  const callToAction = useMemo(
    () =>
      userRole === USER_ROLE_GUEST ? null : (
        <Button onClick={resetGame}>New Game</Button>
      ),
    [resetGame, userRole]
  );

  return (
    <>
      <Header>{header}</Header>
      <Subheader>{subheader}</Subheader>
      {callToAction}
    </>
  );
};

export default GameOver;
