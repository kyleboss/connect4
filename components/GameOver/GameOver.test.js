import React from "react";
import { render } from "@testing-library/react";

import GameOver from "./GameOver";
import useGame from "../../hooks/useGame";
import {
  USER_ROLE_GUEST,
  USER_ROLE_RED,
  USER_ROLE_YELLOW,
} from "../../utils/constants";

jest.mock("../../hooks/useGame", () => ({
  __esModule: true,
  default: jest.fn(),
}));

describe("GameOver", () => {
  it("renders correctly when the user wins", () => {
    useGame.mockImplementation(() => ({
      userRole: USER_ROLE_RED,
      winner: USER_ROLE_RED,
      resetGame: jest.fn(),
    }));

    const { asFragment, getByText } = render(<GameOver />);
    expect(getByText("Congratulations! ðŸŽ‰")).toBeInTheDocument();
    expect(getByText("You won.")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders correctly when the user loses", () => {
    useGame.mockImplementation(() => ({
      userRole: USER_ROLE_YELLOW,
      winner: USER_ROLE_RED,
      resetGame: jest.fn(),
    }));

    const { asFragment, getByText } = render(<GameOver />);
    expect(getByText("Game over.")).toBeInTheDocument();
    expect(getByText("Red won.")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("does not render the new game button for guests", () => {
    useGame.mockImplementation(() => ({
      userRole: USER_ROLE_GUEST,
      winner: USER_ROLE_RED,
      resetGame: jest.fn(),
    }));

    const { queryByText } = render(<GameOver />);
    expect(queryByText("New Game")).not.toBeInTheDocument();
  });
});
