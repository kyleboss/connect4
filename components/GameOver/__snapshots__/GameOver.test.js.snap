// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameOver renders correctly when the user loses 1`] = `
<DocumentFragment>
  <h1
    class="sc-f9cd8f8e-0 dcLxvF"
  >
    Game over.
  </h1>
  <h2
    class="sc-773df940-0 icsDSx"
  >
    Red won.
  </h2>
  <button
    class="sc-5e2d1029-0 NWVxU"
  >
    New Game
  </button>
</DocumentFragment>
`;

exports[`GameOver renders correctly when the user wins 1`] = `
<DocumentFragment>
  <h1
    class="sc-f9cd8f8e-0 dcLxvF"
  >
    Congratulations! ðŸŽ‰
  </h1>
  <h2
    class="sc-773df940-0 icsDSx"
  >
    You won.
  </h2>
  <button
    class="sc-5e2d1029-0 NWVxU"
  >
    New Game
  </button>
</DocumentFragment>
`;
